[p2p] Reduce addr blackholes

amitiuttarwar opened the pull request:
This PR builds on the test refactors extracted into #22306 (first 5 commits). 

This PR aims to reduce addr blackholes. When we receive an `addr` message that contains 10 or less addresses, we forward them to 1-2 peers. This is the main technique we use for self advertisements, so sending to peers that wouldn't relay would effectively "blackhole" the trickle. Although we cannot prevent this in a malicious case, we can improve it for the normal, honest cases, and reduce the overall likelihood of occurrence. Two known cases where peers would not participate in addr relay are if they have connected to you as a block-relay-only connection, or if they are a light client. 

This implementation defers initialization of `m_addr_known` until it is needed, then uses its presence to decide if the peer is participating in addr relay. For outbound (not block-relay-only) peers, we initialize the filter before sending the initial self announcement when processing their `version` message. For inbound peers, we initialize the filter if/when we get an addr related message (`ADDR`, `ADDRV2`, `GETADDR`). We do NOT initialize the filter based on a `SENDADDRV2` message. 

To communicate about these changes beyond bitcoin core & to (try to) ensure that no other software would be disrupted, I have: 
- Posted to the [mailing list](https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2021-April/018784.html)
- Researched other open source clients to confirm compatibility, opened issues in all the projects & documented in https://github.com/bitcoin/bitcoin/pull/21528#issuecomment-809906430. Many have confirmed that this change would not be problematic.
- Raised as topic during [bitcoin-core-dev meeting](https://www.erisian.com.au/bitcoin-core-dev/log-2021-03-25.html#l-954)
- Raised as topic during [bitcoin p2p meeting](https://www.erisian.com.au/bitcoin-core-dev/log-2021-04-20.html#l-439)

amitiuttarwar committed:
[net_processing] Introduce SetupAddressRelay

Idempotent function that initializes m_addr_known for connections that support
address relay (anything other than block-relay-only). Unused until the next
commit.

amitiuttarwar committed:
[net_processing] Defer initializing m_addr_known

Use SetupAddressRelay to only initialize `m_addr_known` as needed. For outbound
peers, we initialize the filter before sending our self announcement (not
applicable for block-relay-only connections). For inbound peers, we initialize
the filter when we get an addr related message (ADDR, ADDRV2, GETADDR).

These changes intend to mitigate address blackholes. Since an inbound peer has
to send us an addr related message to become eligible as a candidate for addr
relay, this should reduce our likelihood of sending them self-announcements.

ajtowns commented on src/net_processing.cpp:
```diff
@@ -2625,7 +2625,11 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
         }
-        pfrom.m_wants_addrv2 = true;
+
+        if (SetupAddressRelay(pfrom)) {
```
I don't think there's any need to `SetupAddressRelay` here, and doing so would cause you to continue choosing inbounds that are treating you as block-relay-only instead of ignoring them as a black hole.

jnewbery replied:
I don't think it matters much either way. The only peers this affects are inbound block-relay-only connections from v0.21 peers, which (incorrectly) send `sendaddrv2` to us. Inbound block-relay-only peers on v22 won't send `sendaddrv2` because of the change in commit _Stop sending SENDADDRV2 message to block relay only peers_.

amitiuttarwar replied:
ergh, its annoying that v0.21 peers will send us the `sendaddrv2` message, but I'm not sure if this warrants removing this conditional. some thoughts:
- in terms of protocol & logical coherence, I'd expect `sendaddrv2` to indicate desire that the node wants to participate in addr relay
- in this patch, calling `SetupAddressRelay` for the node mimics the behavior of current master. so I think of calling as the default (rather than the other way around)
- in terms of safety, it seems better to have a blackhole than exclude a peer from address relay.   I think... a blackhole means an address, or set of addresses will not propagate as well this time, but excluding a peer from address relay (esp if programmatically / many peers exclude) means that peer could be at a higher risk of being eclipsed? 

I thought I was on the fence, but writing out this reasoning makes me realize I think I'd like to leave as is.

thoughts?

amitiuttarwar replied:
Leaving this as is - I find it most logically consistent for `sendaddrv2` to indicate "interested in addresses". Opened #22245 to change the behavior for inbound block-relay-only v22.0 peers.

jnewbery commented on src/net.cpp:
```diff
@@ -611,6 +611,8 @@ void CNode::copyStats(CNodeStats &stats, const std::vector<bool> &m_asmap)
 
     X(m_conn_type);
+
+    stats.addr_relay_enabled = RelayAddrsWithConn();
```
This isn't thread safe. By accessing `m_addr_known` from outside the message handler thread (eg an rpc worker thread or qt), this opens race conditions where `m_addr_known` is being set by the message handler thread and read by an rpc thread concurrently. In such cases, the thread reading the value could see the old value, the new value or garbage.

amitiuttarwar replied:
ayo, good catch. it feels like overkill to introduce a lock just to expose as a boolean, so I've dropped this commit for now. will some of your addr-in-net-processing refactor will make it so we can do this safely?

jnewbery replied:
> will some of your addr-in-net-processing refactor will make it so we can do this safely?

Hopefully, yes!

jnewbery commented on src/net_processing.cpp:
```diff
@@ -4243,12 +4279,17 @@ bool PeerManagerImpl::SendMessages(CNode* pto)
             pto->vAddrToSend.clear();
-            if (!vAddr.empty())
+
+            if (!vAddr.empty()) {
```
I suggest you drop these tidy-up changes in _Add some documentation and small style fixes_. They don't seem very related to the rest of the PR :see_no_evil:

amitiuttarwar replied:
sorry I'm not sorry!! I abhor two line if statements, they are great footguns. these changes are pretty trivial, I don't think they would make sense as a stand-alone PR, or that they increase the complexity of review, so I'd rather leave them in.

jnewbery replied:
I agree that two line if statements are undesirable. If you were already touching these lines, then I'd agree that you should improve them. However, these are the only changes you make to `SendMessages()` and are unrelated to the rest of the changes in this PR.

I refactor this entire chunk of logic in #21236. Without this tidy-up commit, the two PRs don't conflict and can be merged in either order. If you include these changes, then the PRs conflict and one will have to be rebased when the other is merged.

amitiuttarwar replied:
ok! reverted :)

jnewbery reviewed (commented):
Concept ACK. Thanks for implementing this!

You've introduced a tiny thread safety issue by accessing `m_addr_known` from multiple threads without a lock. Y'know, this'd be much tidier if the addr data was all contained in net_processing :grin:

jnewbery commented:
I think that an approach like this was first suggested by AJ in https://github.com/bitcoin/bitcoin/pull/15759#issuecomment-527012757:

> I think since we AdvertiseLocal regularly to our non-block-relay peers, we could just set a flag on the peer if we've ever seen an ADDR message from them, and only choose nodes with that flag in RelayAddress?

I think it's a good idea, since it stops us relaying addrs to both inbound block-relay-only peers _and_ other kinds of black holes.

amitiuttarwar commented:
fixed tests & added a test. patch is ready, but leaving as a draft until I gain more confidence that this would not break any other software on the network.

ccdle12 commented on src/net_processing.cpp:
```diff
@@ -2437,6 +2437,8 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
         }
 
         if (!pfrom.IsInboundConn() && !pfrom.IsBlockOnlyConn()) {
```
nit: is it worth replacing `if (... && !pfrom.IsBlockOnlyConn())` with `if (... && SetupAddrssRelay(pfrom))` since it already checks if the peer is block-relay-only?

ccdle12 commented on test/functional/p2p_addr_relay.py:
```diff
@@ -101,6 +115,25 @@ def relay_tests(self):
 
         self.nodes[0].disconnect_p2ps()
 
+    def blackhole_tests(self):
```
nice tests, would it would be overkill to also test for the outbound case?

amitiuttarwar replied:
definitely not overkill, the inbound vs outbound behavior is different- inbounds we wait to hear an `addr` message before considering the peer eligible for relay, but with outbounds we setup relay, send `getaddr` & self-announce (unless its a block-relay-only connection). I added some tests to check these behaviors.

glozow reviewed (commented):
Concept ACK to making an effort not to send addrs to peers that will likely ignore them. By "disrupt other software on the network," is the concern that some nodes who aren't immediately active in addr-related dialogue might be skipped over and not hear about addrs?

At first I was confused because I figured a lot of nodes don't care about advertising their own addr (e.g. bootstrapping, in IBD, maybe not accepting inbounds). But it seems like a node making a full relay outbound will always send its own addr and a getaddr, and should thus immediately be a candidate for getting addrs forwarded to it (is this accurate?). Was there a version of Bitcoin Core that didn't send addr stuff after getting `VERSION` from outbound?

amitiuttarwar commented on test/functional/p2p_addr_relay.py:
```diff
@@ -82,6 +130,49 @@ def run_test(self):
 
+        self.nodes[0].disconnect_p2ps()
+
+    def inbound_blackhole_tests(self):
+        self.log.info('Check that we only relay addresses to inbound peers who have previously sent us addr related messages')
+
+        addr_source = self.nodes[0].add_p2p_connection(P2PInterface())
+        receiver_peer = self.nodes[0].add_p2p_connection(AddrReceiver())
+        blackhole_peer = self.nodes[0].add_p2p_connection(AddrBlackhole())
+
+        msg = self.setup_addr_msg(2)
+        self.send_addr_msg(addr_source, msg, [receiver_peer, blackhole_peer])
+
+        assert_equal(receiver_peer.num_ipv4_received, 2)
+        assert_equal(blackhole_peer.num_ipv4_received, 0)
+
+        self.nodes[0].disconnect_p2ps()
```
> Since peers are chosen randomly..

yes, peers are chosen randomly, but this test is set up to not be random- ipv4 addresses will be relayed to 2 peers that are not the source of the addr message. so both `receiver_peer` and `blackhole_peer` would receive the addr if they were eligible. definitely wouldn't want a flaky test there :) 

> I think this test could also be improved.. 

thats a great idea! I've added locally, will include in next push. 

> assert bytes received for addr-related messages = 0

hm, `blackhole_peer` is a p2p connection, so we can't call `getpeerinfo()` on it to get the bytes received directly. that said, we could check `bytessent_per_msg` on the node to see that no bytes have been sent for an `addr` message. however, I its a bit brittle & the p2p connection is capturing directly receiving the addr message, so I don't think that's adding much value to the test. let me know if you think it'd add test coverage :)

glozow replied:
> yes, peers are chosen randomly, but this test is set up to not be random- ipv4 addresses will be relayed to 2 peers that are not the source of the addr message. so both receiver_peer and blackhole_peer would receive the addr if they were eligible. definitely wouldn't want a flaky test there :)

Oh right it's always going to be the same one. That makes sense then, that the node has 2 options but doesn't send it to a peer that hasn't established addr relay. But wait I'm confused why it would be flaky, I thought `blackhole_peer` would never get the addr because it hasn't sent any addr-related messages?

> hm, blackhole_peer is a p2p connection, so we can't call getpeerinfo() on it to get the bytes received directly.

Er I mean, call `getpeerinfo()` on the node to see bytes received, kinda like this:

```py
def sum_addr_messages(msgs):
    return msgs['addr'] + msgs['addrv2'] + msgs['getaddr'] + msgs['sendaddrv2']
...

peerinfo = self.nodes[0].getpeerinfo()
# receiver_peer sent us its own addr and a getaddr after the VERSION message, so we relay
assert sum_addr_messages(peerinfo[1]['bytesrecv_per_msg']) > 0
assert_equal(receiver_peer.num_ipv4_received, 2)

# blackhole_peer should not have sent us any addr-related messages, so we don't relay
assert sum_addr_messages(peerinfo[2]['bytesrecv_per_msg']) == 0
assert_equal(blackhole_peer.num_ipv4_received, 0)
```
I just think it's a more precise test, asserting that the addrs aren't relayed because no addr messages have been received. If addr dialogue changes, this test catches it/debugs itself. Not sure if that counts as coverage though

amitiuttarwar replied:
> But wait I'm confused why it would be flaky, I thought blackhole_peer would never get the addr because it hasn't sent any addr-related messages?

yes yes, it would be flaky if it wasn't deterministic / `blackhole_peer` was being skipped by chance. I think we're on the same page, just confusion of my language now. 

> Er I mean, call getpeerinfo() on the node to see bytes received, kinda like this

oh I get it now, you're suggesting additional coverage to test the setup, not another way to test the `addr` relayed. that's a cool idea! sure, I'll add it :) 

smol note: 
> # receiver_peer sent us its own addr and a getaddr after the VERSION message, so we relay

since `receiver_peer` is essentially a `P2PInterface` object, it actually won't self announce upon receiving `version` message. I added  that it will send out a `getaddr` so tests wouldn't unexpectedly break  in this patch. but I think that highlights why this suggested coverage is useful, at the very least it communicates the relevant setup that's otherwise "invisible" from the POV of this test file.

amitiuttarwar replied:
based on this convo, I've updated the tests to assert that `blackhole_peer` starts receiving addr messages after sending one. I've also added `getpeerinfo` for the node to verify the test setup of which peers have sent addr related messages. 

thanks!

sipa commented:
@glozow The concern would be other Bitcoin node implementations which don't send addr/getaddr, but do care about receiving them. I don't think that's particularly likely, but it's also hard to be sure there are none.

ivanacostarubio commented:
Concept ACK

GeneFerneau commented on src/net.h:
```diff
@@ -546,6 +543,20 @@ class CNode
+     *  is if they connected to us as a block-relay-only peer or as a light
+     *  client.
+     * */
     std::unique_ptr<CRollingBloomFilter> m_addr_known{nullptr};
```
```suggestion
    std::unique_ptr<CRollingBloomFilter> m_addr_known GUARDED_BY(cs_addrKnown){nullptr};
```
and add `cs_addrKnown` as a private member:

```
mutable RecursiveMutex cs_addrKnown;
```

to address the point about race conditions mentioned by @jnewbery

amitiuttarwar replied:
Holding off on this to try to figure out if there's a better way I can expose this value through the `getpeerinfo` RPC than having to introduce a mutex for this one hyper-specific use case. Ideas welcome!

amitiuttarwar replied:
added `addr_relay_enabled` as an atomic to address this issue.

GeneFerneau commented:
Concept ACK

Need to review surrounding code more, and run the tests. Looks good so far!

jnewbery commented:
Reiterating my concept ACK on this. Reducing the number of addr black holes seems like a good win for addr gossipping across the network.

This branch needs rebase now that the addr data has been moved up into net_processing. I'd be very happy to help with that if you need a hand.

ajtowns commented:
> Reiterating my concept ACK on this. Reducing the number of addr black holes seems like a good win for addr gossipping across the network.

I think this was somewhat pending being sure it wouldn't break bitcore (see https://github.com/bitpay/bitcore/issues/3140) which now looks like it's now resolved.

amitiuttarwar commented:
> Reiterating my concept ACK on this

Thanks! 

> I think this was somewhat pending being sure it wouldn't break bitcore (see bitpay/bitcore#3140) which now looks like it's now resolved.

Indeed. I'm now working to rebase & bring this PR out of draft. I've rebased locally but am working through some test failures, I hope to have this ready for review by early next week. 

PS: I also investigated bitcoinj & opened an issue on their repo, they also seem compatible 🎉.

amitiuttarwar committed:
[test] Test that we intentionally select addr relay peers.

This test checks that we only relay addresses with inbound peers who have sent
us an addr related message. Uses a combination of GETADDR and ADDR to verify
when peers are eligible.

amitiuttarwar commented on src/net.h:
```diff
@@ -546,6 +543,20 @@ class CNode
+     *  are unlikely to forward them, effectively blackholing self
+     *  announcements. Reasons peers might not support addr relay on the link
+     *  is if they connected to us as a block-relay-only peer or as a light
+     *  client.
```
updated language

amitiuttarwar commented:
My [research](https://github.com/bitcoin/bitcoin/pull/21528#issuecomment-809906430) indicates that these changes would not break any clients, I've opened issues trying to confirm my understand on all the repos, and I have not heard any opposition from the mailing list or on IRC. This seems like the extent of what I can do to make sure these changes are compatible with alternative clients 😛

I've also broken out the commit to "stop sending sendaddrv2 messages to block-relay-only peers" into #22245. 

So I'm bringing this PR out of draft. I have rebased it on top of #21236 and #21707. They were both pretty involved, so I recommend reviewing from scratch. The `p2p_addr_relay.py` functional test is getting pretty big & hits a fair amount of behind-the-scenes p2p mechanisms. I tried to improve the clarity & document relevant mechanisms, but open to suggestions if anything is still unclear.

michaelfolkson commented:
Concept ACK. Impressively thorough investigation of impact on alternative clients :)

vasild commented:
> For inbound peers, we initialize the filter if/when we get an addr related message (ADDR, ADDRV2, GETADDR, SENDADDRV2).

So it may happen that a light client who does not participate in address relay connects to us and sends us `getaddr` in order to learn some peers and we (mis)interpret that and start sending unrequested addresses to that peer in the hopes of further relay. Black hole again.

Or a peer that does not participate in address relay replies to our `getaddr` with `addr` and we (mis)interpret that as "the peer participates in address relay"?

Do we have any estimates of how severe the black holes problem is? I think that currently address relay works, so maybe it is not a "big" problem, or not a problem at all?

What if we relay to 3 instead of 2 random peers? Has that been considered? I would not require protocol changes. Would it solve the black holes problem or diminish it to such extend that it can be ignored?

amitiuttarwar commented:
@vasild

> So it may happen that a light client who does not participate in address relay connects to us and sends us getaddr in order to learn some peers and we (mis)interpret that and start sending unrequested addresses to that peer in the hopes of further relay. Black hole again.
Or a peer that does not participate in address relay replies to our getaddr with addr and we (mis)interpret that as "the peer participates in address relay"?

I believe this question is addressed in the OP- "Although we cannot prevent this in a malicious case, we can improve it for the normal, honest cases, and reduce the overall likelihood of occurrence." 

Incase that is unclear, let me try to provide some additional context- the goal of this PR is not to ensure perfect accuracy of whether or not the receiving peer would forward the addresses. In fact, I don't think we could ever ensure such accuracy- even if we had a p2p message that said "I will forward addresses you send me", that could easily be abused by a peer who doesn't actually follow through. 

The goal of this PR is to _reduce the frequency_ of how frequently we send addresses who have no interest in them & will certainly not forward those along. For example, we know that an honest node who is running the latest version of bitcoin core will have block-relay-only connections where it will drop any addresses sent over that connection.

> Do we have any estimates of how severe the black holes problem is? I think that currently address relay works, so maybe it is not a "big" problem, or not a problem at all? 

While I do think this PR offers an improvement to address relay, the end goal is to increase the number of block-relay-only connections created by a bitcoin node (also goal of the `DISABLETX` work in #20726). Even if the number of black holes are not currently a problem, fixing them are a prerequisite to being able to increase the number of block-relay-only connections. For more context on this conversation, please see https://github.com/bitcoin/bitcoin/pull/15759#issuecomment-480868516 & the surrounding discussion. 

> What if we relay to 3 instead of 2 random peers? Has that been considered? I would not require protocol changes. Would it solve the black holes problem or diminish it to such extend that it can be ignored?

Relaying addresses to 3 rather than 2 random peers could be possible, but I think it would require in-depth research to ensure there wouldn't be any negative bandwidth or privacy impact. One of the alternative approaches that has been considered is to mark "potential blackholes" and then increase redundancy if they have been selected in `RelayAddress`. A downside of that approach is it's hard to build confidence that such approach wouldn't lead to network-wide biases of certain addresses or certain peers. 

I agree with what @sdaftuar said in https://github.com/bitcoin/bitcoin/pull/22245#issuecomment-862839610: 
> I don’t think anyone believes that implicit behavior (that is undocumented in any BIP) is the right way for protocol development to proceed if starting from a clean slate — so eventually I think we’ll rewrite things.

The current state of `addr` propagation relies heavily on assumptions & is poorly understood and poorly documented. In the long term, I hope to work towards a shared set of design goals & more explicit expectations that are well documented. However, this PR is coming from the hope that we don't have to gate making progress towards increasing block-relay-only connections on a fundamental redesign of `addr` relay.

vasild commented:
If the focus is on avoiding gossip to peers that connected to us in block-only mode, then isn't the absence of `getaddr` enough to recognize that? That is, a peer who connected to us and did not send us `getaddr` must be in blocks-only mode / is a black hole? No need to involve `addr` and `addrv2`? I.e. only for inbound connections, delay initializing `Peer::m_addr_known` for until after `getaddr` is received.

Given that "when" and "to whom" to relay addresses is not mandated by any specification and is up to each peer to decide, and one may as well decide to relay only on full moon on odd months or if the peer's address ends in `7`, then deciding to skip relay to peers that did not send `getaddr` is not a protocol change.

I see the assumption "the peer did not send us `messageX` => they must be a black hole" as flawed. More like a hack, which might be acceptable in the current situation.

Still sending to some black holes is ok (we currently do). But could this PR have the adverse effect of starting to treat as black holes peers that are actually not?

As for relaying to more peers: currently we relay reachable addresses to 2 peers. So (assuming everybody relays) every node relays to two others (it spreads like a virus). And an average spread rate of <1 is required to stop an address from propagating. It means that if more than half of the peers are black holes then propagation will not work (will stop). I think we are far from 50% black holes but I could be wrong. For unreachable peers (e.g. a Tor address relayed via a peer that does not have Tor connectivity) the number is 1.5 and so, 33% or more black holes are needed to brick propagation.

amitiuttarwar commented:
@vasild 

> That is, a peer who connected to us and did not send us getaddr must be in blocks-only mode / is a black hole?

This is false. If you see the notes I've taken on alternative clients here: https://github.com/bitcoin/bitcoin/pull/21528#issuecomment-809906430, you can see that several of them will always self-advertise, but only occasionally send getaddr messages. 

> Still sending to some black holes is ok (we currently do). But could this PR have the adverse effect of starting to treat as black holes peers that are actually not?

Yes, exactly. This is why I have taken great care to communicate these changes & research clients to ensure they would not be harmed. If we only used getaddr to indicate "wants to participate in addr relay", this exact adverse effect would occur. 

> As for relaying to more peers: currently we relay reachable addresses to 2 peers. So (assuming everybody relays) every node relays to two others (it spreads like a virus). And an average spread rate of <1 is required to stop an address from propagating. It means that if more than half of the peers are black holes then propagation will not work (will stop). I think we are far from 50% black holes but I could be wrong. For unreachable peers (e.g. a Tor address relayed via a peer that does not have Tor connectivity) the number is 1.5 and so, 33% or more black holes are needed to brick propagation.

As I said in my previous post, "Relaying addresses to 3 rather than 2 random peers could be possible, but I think it would require in-depth research to ensure there wouldn't be any negative bandwidth or privacy impact." There is no doubt that relaying addresses to more peers could ensure they saturate the network. Heck, if that was all we cared about, why not forward addresses to every node? 

The concern is if there would be any _other_ adverse impact, and that is difficult to reason about. We could model the bandwidth increase by estimating: number of nodes on network, expected frequency of initiating self-announcements, number of anticipated blackholes & multiplier of forwarding addresses. This could help build confidence, but a lot of these numbers would have to be estimates. Even harder would be to guarantee that there are no significant security concerns. For example, addr relay can reveal node topology which can ease a partition or eclipse attack. I believe it would be significantly more complex to usefully model how changing the relay policy would impact potential information leaks. And that wouldn't be the only security concern about changing addr relay policy, just one example.

mzumsande commented:
Concept ACK.

> As for relaying to more peers: currently we relay reachable addresses to 2 peers. So (assuming everybody relays) every node relays to two others (it spreads like a virus). And an average spread rate of <1 is required to stop an address from propagating. It means that if more than half of the peers are black holes then propagation will not work (will stop). I think we are far from 50% black holes but I could be wrong.

I think that there are at least two limitations to a virus-like spread:
1. Each full node relays a given addr only once to two peers, and becomes a black hole for it after that - if it receives the addr again from another peer, it won't relay it to different peers within 24h because of how `RelayAddress()` works. This makes addr gossip sensitive to network topology (hubs with many peers are bottlenecks) and makes it pretty much impossible for an addr to percolate through the entire network via gossip relay. Even with zero black holes, an addr would only reach ~35% of the nodes in my simulation.
2. Each node in the network will stop relaying an addr 10 minutes after the announcement (that's 20 hops on average with `AVG_ADDRESS_BROADCAST_INTERVAL = 30s`), see [first condition here](https://github.com/bitcoin/bitcoin/blob/da69d9965a112c6421fce5649b5a18beb7513526/src/net_processing.cpp#L2783-L2786)

My impression is that the introduction of 20% block-relay-only connections could have effected addr gossip relay quite clearly. I did some quick simulations similar to @naumenkogs' work from #17194 but adjusted for points 1 and 2 from above: 
The simulations ([code](https://github.com/mzumsande/addr_relay_sim)) suggest that the percentage of nodes reached by an initial self-announcement of a new non-listening node falls from 35% (0% black holes) to 21% (20% black holes) to 2% (35% black holes).

> I think that currently address relay works, so maybe it is not a "big" problem, or not a problem at all?

I think that whether addr gossip relay "works"  is not so easy to judge because other means of addr propagation (DNS seeds; getaddr mechanism) also contribute to addr propagation and introduce some degree of redundancy - maybe the best measure is how fast a new node on the network receives inbound connections?
But getaddr does not adapt quickly (older addrs, 24 hour cache) to changes in the network, so it makes sense to me to improve the state of gossip relay as suggested by this PR.

ajtowns commented:
> The simulations ([code](https://github.com/mzumsande/addr_relay_sim)) suggest that the percentage of nodes reached by an initial self-announcement of a new non-listening node falls from 35% (0% black holes) to 21% (20% black holes) to 2% (35% black holes).

I think that's simulating having 8 outbounds of which 20% (or 30%,35%,40%) are blackholes, rather than 8 full outbounds and an additional 2 blackholes.

Might be worth distinguishing between the percentage of listening nodes that see an address versus total nodes -- I think with 8 outbounds, 0% blackholes it's 92% of listening nodes and 35% of all nodes that see an address, and 20% blackholes (6.4 full outbound 1.6 block relay only) gives 65% of listening nodes and 21% of all nodes see an address.

I tried writing [my own sim](https://gist.github.com/ajtowns/ec27774113e5d1214b99281f6b7549c4) which only deals with listening peers; the numbers I got were 8 full-relay 0 block-relay gives ~80% of nodes seeing an address on average, while 8 full-relay 2 block-relay gives ~70% coverage, and 8 full-relay 5 block-relay gives ~50% coverage. Switching from 2 peers to 3 for addr relay brings those numbers up to 95%, 92% and 85% respectively.

vasild commented:
@mzumsande, @ajtowns, wow! That gives some rough picture on the severity of the problem. I think it rules out "not a problem at all". Thanks!

@amitiuttarwar, I guess the "waiting for confirmation" for btcd and bitcoinj [above](https://github.com/bitcoin/bitcoin/pull/21528#issuecomment-809906430) should be cleared before this PR is merged?

amitiuttarwar committed:
[test] Update p2p_addr_relay test to prepare

Use an init param to make clear whether a getaddr message should be sent when
the P2PConnection receives a version message. These changes are in preparation
for upcoming commits that modify the behavior of a bitcoind node and the test
framework.

amitiuttarwar commented:
@mzumsande, thanks for putting together that awesome simulation! and thank you @ajtowns for reviewing :) very cool. 

biggest changes in the recent push: 
- as per the conversation on #22245, I've removed the commit that uses `sendaddrv2` as a signal to set up addr relay on the connection 
- I was finding the `p2p_addr_relay.py` refactor commit hard to parse, so I pulled it out into #22306 and broke it down into multiple commits. this PR builds on that branch.

ajtowns commented:
Had a bit more of a [poke](https://gist.github.com/ajtowns/ec27774113e5d1214b99281f6b7549c4) at this, I think the interesting results are (with 10k listening nodes, 50k private nodes with only outbounds):

```
Conn: 8+2 Spread: 2 Limit: 10000/600 Avg/dev: 78.5%±9.7% 26.1%±3.5%
Conn: 8+0 Spread: 2 Limit: 10000/600 Avg/dev: 90.5%±1.7% 33.5%±1.0%
Conn: 8+5 Spread: 2 Limit: 10000/600 Avg/dev: 44.5%±17.0% 13.0%±5.1%

Conn: 8+5 Spread: 3 Limit: 10000/600 Avg/dev: 96.5%±9.7% 35.8%±3.6%
Conn: 8+0 Spread: 3 Limit: 10000/600 Avg/dev: 99.6%±0.1% 46.1%±0.2%
```

That is, with 8 regular outbounds, sending addresses to 2 peers gives ~78% coverage of listening peers (and ~26% coverage of total peers) with 2 additional block-relay-only/blackhole peers, or ~90% coverage with this patch, or ~44% coverage without this patch and 5 additional block-relay-only peers instead of 2.

With a spread of 3 instead, then 5 block-relay-only peers gets to 96% coverage even without this patch (though with a large stddev), and with this patch as well gets to 99% coverage with a small standard deviation. So this seems like a "both" rather than an "either/or" situation to me.

Addr relay is pretty low bandwidth, so I think the extra 50% of data for going from 2 to 3 is probably pretty fine (maybe make it 3 peers of which at most 2 are outbounds so that private peers don't increase traffic to listening nodes?); I'm not seeing what the privacy concerns should really be.

I think these numbers are a bit more robust and better match @mzumsande's results. I switched the limit to 600s with poisson timing rather than 20 hops; provided there were enough samples the results between the two methods seemed pretty consistent.

So in summary, Concept ACK?

vasild commented:
> Addr relay is pretty low bandwidth, so I think the extra 50% of data for going from 2 to 3 is probably pretty fine...

I agree.

> I'm not seeing what the privacy concerns should really be.

Me neither.

I reproduced the above results and ran one more simulation: just increase spread from 2 to 3:

```
# current (copy from above)
Conn: 8+2 Spread: 2 Limit: 10000/600 Avg/dev: 78.5%±9.7% 26.1%±3.5%

# just increase spread from 2 to 3
Conn: 8+2 Spread: 3 Limit: 10000/600 Avg/dev: 99.1%/0.1% 41.4%/0.2%

# increase spread from 2 to 3 and reduce blackholes as per this PR (copy from above)
Conn: 8+0 Spread: 3 Limit: 10000/600 Avg/dev: 99.6%±0.1% 46.1%±0.2%
```

So, assuming spread is increased from 2 to 3 and we are at 99.1%/41.4% then this PR will only improve it to 99.6%/46.1% :eyes: 

Also to consider relay of addresses from non-reachable networks, currently with spread 1.5. The script gives an error when given a float: TypeError: can't multiply sequence by non-int of type 'float'.

ajtowns commented:
> Also to consider relay of addresses from non-reachable networks, currently with spread 1.5.

Changing the spread to `random.choice([1,2])` to simulate an address everyone considers unreachable gives about 5% coverage with 8 outbounds and 2 blackholes; but 27% with 8 outbounds and no blackholes (ie behaviour prior to block relay only nods or after this pr). If you wanted to model ipv4 vs ipv6 or tor, you'd need to model some nodes as considering those nets reachable and being peered with other nodes on the network (preferentially?) which would get complicated.

amitiuttarwar committed:
[test] Use the new endpoint to improve tests

amitiuttarwar committed:
[net_processing] Introduce new field to indicate if addr relay is enabled

amitiuttarwar committed:
[net_processing] Remove RelayAddrsWithPeer function

Now that we have a simple boolean stored on the field, the wrapper function is
no longer necessary.

amitiuttarwar committed:
[RPC] Add field to getpeerinfo to indicate if addr relay is enabled

amitiuttarwar commented:
### Latest push: 
- rebased on master 
- introduced a new variable, `addr_relay_enabled`, to the Peer struct. this is used to expose this information to `getpeerinfo` without causing a thread safety issue. The bool made the `RelayAddrsWithPeer` function redundant, so I removed that.
- these commits are generally small & isolated, so I'm hoping this PR is pretty straightforward to review :) 

### Status of this PR: 
- Ready for review! 🔎
- My research of alternative clients suggests these changes would not harm other software: https://github.com/bitcoin/bitcoin/pull/21528#issuecomment-809906430
- I believe I have addressed all of the conceptual concerns that have been expressed by reviewers.  

### Future thought: 

If this PR gets accepted & released, we should keep an eye on the peer rate limiting introduced in #22387. `MAX_ADDR_RATE_PER_SECOND` is currently set to `0.1` based on network observations + some buffer. If this patch were widely adopted, it's possible the observed rate could naturally increase. 

☝️ Further explanation: based on an estimated number of listening nodes on the network & frequency of self announcements in Bitcoin Core, I was surprised to find the observed addr announcement rate / per / second to be ~0.025. I think a large part of the delta between conceptual math & observed frequency arises from blackhole rate (aka how well does a single announcement propagate) combined with the role of `m_addr_known`, which would fluctuate based on the node-path a specific announcement takes. 

Anyway, just want to mention that since this is a network level phenomena, its hard for us to accurately predict, and thus would want to observe.

jnewbery commented on test/functional/p2p_addr_relay.py:
```diff
@@ -184,7 +198,68 @@ def relay_tests(self):
         self.nodes[0].disconnect_p2ps()
 
+    def sum_addr_messages(self, msgs_dict):
+        val = 0
+        for msg in ['addr', 'addrv2', 'getaddr', 'sendaddrv2']:
+            if msg in msgs_dict:
+                val += msgs_dict[msg]
+        return val
```
```suggestion
        return sum(bytes_recvd for (msg, bytes_recvd) in msgs_dict.items() if msg in ['addr', 'addrv2', 'getaddr', 'sendaddrv2'])
```

jnewbery replied:
Also, remove `sendaddrv2` from the list?

amitiuttarwar replied:
done and done

jnewbery commented on src/net_processing.cpp:
```diff
@@ -3725,6 +3743,8 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
         }
         peer->m_getaddr_recvd = true;
 
+        SetupAddressRelay(pfrom);
```
Perhaps move this above the `m_getaddr_recvd` logic (to avoid reading the unused `m_getaddr_recvd` value for non-addr-relay peers).

amitiuttarwar replied:
I don't follow- there is an earlier clause that returns early if `!IsInboundConn`, so `SetupAddressRelay` should never return false. So I don't think the order matters here?

amitiuttarwar replied:
ah gotcha, ok! moved

jnewbery commented on src/net_processing.cpp:
```diff
@@ -2572,7 +2587,8 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
         }
 
-        if (!pfrom.IsInboundConn() && !pfrom.IsBlockOnlyConn()) {
+        // Self advertisement logic
```
This is a bit misleading. The logic block below contains self-advertisement logic, but it also sends a `getaddr` to request addresses, which is not self-advertisement.

amitiuttarwar replied:
good point, updated to include

jnewbery commented on src/net_processing.cpp:
```diff
@@ -224,9 +224,23 @@ struct Peer {
+     *  address related message (ADDR, ADDRV2, GETADDR).
+     *
+     *  We use the presence of this filter to decide whether a peer is eligible
+     *  for trickle relay of addr messages. This avoids relaying to peers that
```
I'd very much prefer not to introduce the word 'trickle' here. I've only ever heard the word 'trickle' being used to refer to transaction propagation (eg #7125). Adding the word here would be overloading the term with a new meaning.

amitiuttarwar replied:
ah gotcha. I updated the language to use "gossiping addr messages", does that feel more appropriate / consistent?

jnewbery replied:
"gossiping addr messages" SGTM!

jnewbery commented on test/functional/p2p_addr_relay.py:
```diff
@@ -184,7 +198,68 @@ def relay_tests(self):
+        # And that peer did not receive addresses
+        assert_equal(blackhole_peer.num_ipv4_received, 0)
+
+        self.log.info("After blackhole peer sends addr message, it becomes eligible for trickle relay")
```
Same comment as above about the word "trickle"

amitiuttarwar replied:
convo continued in https://github.com/bitcoin/bitcoin/pull/21528#discussion_r677807103, resolving this thread

jnewbery commented on src/net_processing.cpp:
```diff
@@ -224,9 +224,23 @@ struct Peer {
+     *  block-relay-only connections) or when an inbound peer sends us an
+     *  address related message (ADDR, ADDRV2, GETADDR).
+     *
+     *  We use the presence of this filter to decide whether a peer is eligible
+     *  for trickle relay of addr messages. This avoids relaying to peers that
+     *  are unlikely to forward them, effectively blackholing self
+     *  announcements. Reasons peers might not support addr relay on the link
+     *  include that they connected to us as a block-relay-only peer or they
+     *  are a light client.
```
This is no longer true. We use `addr_relay_enabled` to decide whether a peer is eligible for addr relay.

amitiuttarwar replied:
ah good point. updated the comments on both `m_addr_known` and `m_addr_relay_enabled.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -224,9 +224,23 @@ struct Peer {
+    std::unique_ptr<CRollingBloomFilter> m_addr_known;
+    /** Whether we are participating in address relay with this connection.
+     *  Must correlate with whether m_addr_known has been initialized. */
+    std::atomic_bool addr_relay_enabled{false};
```
```suggestion
    std::atomic_bool m_addr_relay_enabled{false};
```

amitiuttarwar replied:
done, thanks

jnewbery commented on src/net_processing.h:
```diff
@@ -31,6 +31,7 @@ struct CNodeStateStats {
     std::vector<int> vHeightInFlight;
     uint64_t m_addr_processed = 0;
     uint64_t m_addr_rate_limited = 0;
+    bool addr_relay_enabled;
```
```suggestion
    bool m_addr_relay_enabled;
```

amitiuttarwar replied:
done

jnewbery commented on src/net_processing.cpp:
```diff
@@ -4423,6 +4443,24 @@ class CompareInvMempoolOrder
+    if (node.IsBlockOnlyConn()) return false;
+
+    PeerRef peer = GetPeerRef(node.GetId());
+    if (!peer->addr_relay_enabled) {
+        // First addr message we have received from the peer, initialize
+        // m_addr_known
+        peer->m_addr_known = std::make_unique<CRollingBloomFilter>(5000, 0.001);
+        peer->addr_relay_enabled = true;
```
You can atomically test and set the flag as follows:

```suggestion
    if (!peer->addr_relay_enabled.exchange(true)) {
        // First addr message we have received from the peer, initialize
        // m_addr_known
        peer->m_addr_known = std::make_unique<CRollingBloomFilter>(5000, 0.001);
```

(`std::atomic<T>::exchange()` returns the value of the object before the call (https://en.cppreference.com/w/cpp/atomic/atomic/exchange))

amitiuttarwar replied:
nice trick! done

jnewbery reviewed (commented):
Looks good.

I think the use of an atomic `addr_relay_enabled` is fine for now. Perhaps once #21527 has landed and we have proper locking in net_processing, `m_addr_known` can be guarded by the net processing lock and we can lose the additional atomic bool.

amitiuttarwar commented on src/net_processing.cpp:
```diff
@@ -1681,7 +1696,7 @@ void PeerManagerImpl::RelayAddress(NodeId originator,
 
     for (auto& [id, peer] : m_peer_map) {
-        if (RelayAddrsWithPeer(*peer) && id != originator && IsAddrCompatible(*peer, addr)) {
+        if ( peer->addr_relay_enabled && id != originator && IsAddrCompatible(*peer, addr)) {
```
fixed

amitiuttarwar commented on test/functional/p2p_addr_relay.py:
```diff
@@ -166,11 +178,13 @@ def relay_tests(self):
         # large GETADDR responses from being relayed, it now typically affects the self-announcement
         # of the outbound peer which is often sent before the GETADDR response.
         assert_equal(inbound_peer.num_ipv4_received, 0)
+        inbound_peer.send_and_ping(msg_getaddr())
```
good idea! done & added a comment

amitiuttarwar commented:
thanks for the review @jnewbery! incorporated all your feedback with one or two questions. 

> I think the use of an atomic addr_relay_enabled is fine for now. Perhaps once #21527 has landed and we have proper locking in net_processing, m_addr_known can be guarded by the net processing lock and we can lose the additional atomic bool.

sounds good, I would definitely prefer that approach.

jnewbery commented on src/net_processing.h:
```diff
@@ -31,6 +31,7 @@ struct CNodeStateStats {
     std::vector<int> vHeightInFlight;
     uint64_t m_addr_processed = 0;
     uint64_t m_addr_rate_limited = 0;
+    bool m_addr_relay_enabled;
```
Maybe explicitly default initialize this to `false`:

```suggestion
    bool m_addr_relay_enabled{false};
```

amitiuttarwar replied:
done

jnewbery reviewed (commented):
Code review ACK ceb20ad575cb2cbe971844e8a6422eca87d44bbb

In thread https://github.com/bitcoin/bitcoin/pull/21528#discussion_r677304728 I've suggested follow-up work where all of the address data members are stored in a sub structure that only gets initialized if we're relaying addresses with the peer. I have an example implementation of that here: https://github.com/jnewbery/bitcoin/tree/pr21528.1 if you're interested.

amitiuttarwar commented on src/net_processing.cpp:
```diff
@@ -612,6 +633,14 @@ class PeerManagerImpl final : public PeerManager
+     *  @return   True if address relay is enabled with peer
+     *            False if address relay is disallowed
+     */
+    bool SetupAddressRelay(CNode& node) const;
```
removed

amitiuttarwar commented:
@jnewbery 
- incorporated all your feedback
- also took a look at your branch, it looks awesome! I think it makes a lot of sense to encapsulate the address related data into a small data structure & only initialize where needed. I'd be happy to see this sorta thing through as a follow-up, if these changes get accepted. thank you!

jnewbery commented:
Code review ACK 00a54cbddbc6bc69c30a9597f6e669b1af331d65

MarcoFalke commented on src/net_processing.cpp:
```diff
@@ -230,8 +230,25 @@ struct Peer {
+     *
+     *  We use this bool to decide whether a peer is eligible for gossiping
+     *  addr messages. This avoids relaying to peers that are unlikely to
+     *  forward them, effectively blaackholing self announcements. Reasons
```
7e4662b8a0ea0428aa9eefe7b8116f85e5eda486: blackholing

amitiuttarwar replied:
*** blaaaaackholing 

:) fixed

jnewbery commented:
Code review ACK a8834fa0513d6bfcd4dfc759bbb841c02fcf8517

amitiuttarwar commented on src/net_processing.cpp:
```diff
@@ -4423,6 +4431,23 @@ class CompareInvMempoolOrder
+    // information of addr traffic to infer the link.
+    if (node.IsBlockOnlyConn()) return false;
+
+    PeerRef peer = GetPeerRef(node.GetId());
```
no particular reason _not_ to, I'm wondering what the difference is... is this suggestion slightly less overhead because `GetPeerRef` would construct an additional `shared_ptr`, which would then have to update the internal state used to coordinate. VS if we just pass in a reference (I'm thinking `Peer&`?), there's less bookkeeping since the lifetime of the reference would just be managed in the normal program execution?

MarcoFalke replied:
Also, you wouldn't need to check for nullptr, as the caller already did that

jnewbery replied:
> another alternative: passing in Peer& peer. does this mean the lifetime of the reference would just be managed in the normal program execution? so would that mean this is the most efficient?

This is the way to do it. The called function is not taking ownership of `Peer`, so just pass by reference. You only need to pass smart pointers when you're indicating ownership semantics: https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rf-inout (and items F26/27)

amitiuttarwar replied:
👍 done, thanks!

amitiuttarwar commented:
thanks @MarcoFalke, addressed your comments

jnewbery commented:
reACK 3f7250b328b8b2f5d63f323702445ac5c989b73d

Only changes are addressing review comments from Marco.

glozow commented:
ACK 3f7250b328b8b2f5d63f323702445ac5c989b73d

Compared the `SetupAddressRelay()` call sites to where I thought they should be, threw in some asserts where I expected `m_addr_relay_enabled` to be true, did a code review looking for possibilities of `m_addr_known` being used before initialized. Test looks correct. Thanks for diligently checking for compatibility with other clients, seems to be a safe change. :)

ajtowns commented on src/net_processing.cpp:
```diff
@@ -3717,6 +3743,8 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
             return;
         }
 
+        SetupAddressRelay(pfrom, *peer);
```
`if (!pfrom.IsInboundConn() || !SetupAddressRelay(pfrom, *peer)) { error; return; }` might have been better so you don't have to keep the knowledge that SetupAddressRelay always returns true for inbound conns in your head.

amitiuttarwar replied:
makes sense, I'll change if I have to push again

MarcoFalke replied:
`Assume(Setup())` would do that?

amitiuttarwar replied:
addresses in #22618

ajtowns reviewed (approved):
utACK 3f7250b328b8b2f5d63f323702445ac5c989b73d

I don't think the potential interaction with #22387 should be an issue -- if the rate limit wasn't being hit before it was because addresses weren't being relayed, and if they're hit now the result is just that addresses aren't being relayed.

fanquake commented on test/functional/p2p_addr_relay.py:
```diff
@@ -167,6 +179,9 @@ def relay_tests(self):
         # of the outbound peer which is often sent before the GETADDR response.
         assert_equal(inbound_peer.num_ipv4_received, 0)
 
+        # Send an empty ADDR message to intialize address relay on this connection.
```
s/intialize/initialize

MarcoFalke commented:
Maybe add a line to the release notes saying that inbound peers will never receive address messages unless they send ADDR, ADDRV2, or GETADDR themselves?

MarcoFalke commented on test/functional/p2p_addr_relay.py:
```diff
@@ -11,14 +11,15 @@
     msg_addr,
-    msg_getaddr
+    msg_getaddr,
+    msg_verack
```
nit: Would be nice to always add a trailing comma to avoid ugly diffs and keeping the `git blame` depth lower in the future.

amitiuttarwar replied:
noted

amitiuttarwar commented:
@MarcoFalke 
> Maybe add a line to the release notes

Added in #22618
